phabricatoremails/logging.py

```diff
@@ -0,0 +1,41 @@
+import logging.config
+
+
+def create_dev_logger():
+    logging.config.dictConfig(
+        {
+            "version": 1,
+            "handlers": {
+                "console": {"level": "DEBUG", "class": "logging.StreamHandler"}
+            },
+            "loggers": {
+                "phabricator-emails": {"handlers": ["console"], "level": "DEBUG"}
+            },
+        }
+    )
+    return logging.getLogger("phabricator-emails")
```

Comment:
`logging.getLogger(...)` returns the logger, which is used by the caller of `create[_dev]_logger()`.
I didn't want to use `logging.getLogger(...)` statically, since:
1. Globals make testing harder :)
2. The logger config we use depends on our settings (`create_dev_logger()`, `create_logger()`)

Reply:
> Technically you are setting the logger on a global level anyway -- but your current approach is obfuscating how you are doing it

That's exactly my intention. Yes, `logging` still creates an manages a `phabricator-emails` logger somewhere in the global namespace. However, the downsides of this approach are _abstracted away_ in the implementation here.

To be clear, the downsides of the "global" `logging` approach is that it's not clear how or where a logger is configured. It obfuscates the dependencies of modules. since they aren't provided loggers when they're used.

If code throughout a project has `logging.getLogger(logger_key)`, it's not clear how and where this logger is configured. You can grep through the project, but there's no way to guarantee that there's no sneaky global modification happening somewhere that you've missed. `logging.getLogger(phabricatoremails.logging.KEY)` is _better_ for this specific issue, since you can look for usages of `phabricatoremails.logging.KEY`, but it's still adding complexity to your codebase.

![getLogger](https://user-images.githubusercontent.com/7784737/79155298-c1934d00-7d85-11ea-86fb-23d0c8ed61cd.png)

Alternatively, by abstracting away the global logger store and distributing our logger explicitly, it's easier to work backwards to discover where a logger comes from and how it's configured. Additionally, it's less complex:

![direct](https://user-images.githubusercontent.com/7784737/79155912-bd1b6400-7d86-11ea-8135-6792847f5cb2.png)

Reply:
How about this approach? Note that I added the `Environment` enum here, but there are other ways to implement the same functionality. I think this is a little cleaner, and factors out the repeated `logging.getLogger`, but keeps your approach of separating the two logger creations.

```python
# at the top of this module

LOGGER_CONFIGS = {
    Environment.DEV: {
        "version": 1,
        "handlers": {
            "console": {"level": "DEBUG", "class": "logging.StreamHandler"}
        },
        "loggers": {
            "phabricator-emails": {"handlers": ["console"], "level": "DEBUG"}
        },
    },
    Environment.PROD: {
        "version": 1,
        "formatters": {
            "json": {
                "()": "dockerflow.logging.JsonLogFormatter",
                "logger_name": "phabricator-emails",
            }
        },
        "handlers": {
            "console": {
                "level": "DEBUG",
                "class": "logging.StreamHandler",
                "formatter": "json",
            }
        },
        "loggers": {
            "phabricator-emails": {"handlers": ["console"], "level": "DEBUG"}
        },
    },
}

# below in this module
def create_logger(env):
    logging.config.dictConfig(LOGGER_CONFIGS[env])
    return logging.getLogger("phabricator-emails")

# in settings.py
def _parse_logger(config: ConfigParser):
    # NOTE: I would also suggest that the environment should have its own
    # definition in the config rather than relying on the existence of a section
    env = Environment.DEV if config.has_section("dev") else Environment.PROD
    return create_logger(env)
```